### PR TITLE
initial version of INI file parser

### PR DESCRIPTION
**Pull request by maximiliankaul**
Initial version of an INI file parser.
- [x] check license compatibility

### CODE REVIEW

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * A simple language representing classical [INI files](https://en.wikipedia.org/wiki/INI_file). As
+ * there are conflicting definitions of an INI file, we go with:
+ * - the file extension is `.ini`
+ * - all entries live in a unique `section`
+ * - all `key`s are unique per section
+ * - the file is accepted by the [ini4j library](https://ini4j.sourceforge.net/)
+ */
+class IniFilesLanguage : Language<IniFilesFrontend>() {
+    override val fileExtensions = listOf("ini")
```
Comment (maximiliankaul) [inline cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt:42]: ".conf" extension, too

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    override val namespaceDelimiter: String = "" // no such thing
```
Comment (maximiliankaul) [inline cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt:43]: path separator `.`

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    override fun typeOf(type: Any?): Type {
+        return primitiveType("string")
+    }
+
+    override fun codeOf(astNode: Any): String? {
+        return astNode.toString()
+    }
+
+    override fun locationOf(astNode: Any): PhysicalLocation? {
+        return null // currently, the line number / column cannot be accessed given an Ini object
```
Comment (maximiliankaul) [inline cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt:122]: return entire file first line number to last line number

File: cpg-language-configfiles/src/test/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesTest.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Test
+    fun gettingStartedWithINIConfigfiles() {
+        val topLevel = Path.of("src", "test", "resources")
+        val tu =
+            analyzeAndGetFirstTU(listOf(topLevel.resolve("config.ini").toFile()), topLevel, true) {
+                it.registerLanguage<IniFilesLanguage>()
+            }
+        assertIs<TranslationUnitDeclaration>(tu)
+
```
Comment (maximiliankaul) [inline cpg-language-configfiles/src/test/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesTest.kt:51]: add namespace test

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt (hunk 1 above)
Comment (maximiliankaul) [reply cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt]: done

File: cpg-language-configfiles/src/test/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesTest.kt (hunk 4 above)
Comment (maximiliankaul) [reply cpg-language-configfiles/src/test/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesTest.kt:53]: done

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt (hunk 3 above)
Comment (maximiliankaul) [reply cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt]: I updated this to return a Region of -1, -1, -1, -1. Seems more reasonable to me.

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt (hunk 2 above)
Comment (maximiliankaul) [reply cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt]: done

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+ * - [codeOf] has to accept `Any` (because of the limitations stated above) and simply returns
+ *   `.toString()`
+ * - [locationOf] always returns `null` as the "ini4j" library does not provide any means of getting
+ *   a location given a node
+ * - [setComment] TODO what's this?
+ */
+class IniFilesFrontend(language: Language<IniFilesFrontend>, ctx: TranslationContext) :
+    LanguageFrontend<Any, Any?>(language, ctx) {
+
+    private lateinit var uri: URI
```
Comment (maximiliankaul) [inline cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt:71]: Is this ok to do? I'm thinking about multiple calls to `parse` in parallel.

File: cpg-language-configfiles/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ *                    \______/ \__|       \______/
+ *
+ */
+plugins {
+    id("cpg.frontend-conventions")
+}
+
+publishing {
+    publications {
+        named<MavenPublication>("cpg-language-configfiles") {
```
Comment (oxisto) [inline cpg-language-configfiles/build.gradle.kts:32]: `cpg-language-configfiles` looks a bit weird for me, maybe `cpg-language-config-files`?

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+ *   type
+ * - [codeOf] has to accept `Any` (because of the limitations stated above) and simply returns
+ *   `.toString()`
+ * - [locationOf] always returns `null` as the "ini4j" library does not provide any means of getting
+ *   a location given a node
+ * - [setComment] not implemented as this is not used (no
+ *   [de.fraunhofer.aisec.cpg.frontends.Handler] pattern implemented)
+ * - Comments in general are not supported.
+ */
+class IniFilesFrontend(language: Language<IniFilesFrontend>, ctx: TranslationContext) :
```
Comment (oxisto) [inline cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt:70]: Should probably be `IniFileFrontend`, we always had a singular in the frontend, also similarly in the language

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * A simple language representing classical [INI files](https://en.wikipedia.org/wiki/INI_file). As
+ * there are conflicting definitions of an INI file, we go with:
+ * - the file extension is `.ini`
+ * - all entries live in a unique `section`
+ * - all `key`s are unique per section
+ * - the file is accepted by the [ini4j library](https://ini4j.sourceforge.net/)
+ */
+class IniFilesLanguage : Language<IniFilesFrontend>() {
```
Comment (oxisto) [inline cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt:41]: `IniFileLanguage`

File: cpg-language-configfiles/src/test/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesTest.kt [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+import java.nio.file.Path
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertIs
+import kotlin.test.assertNotNull
+
+class IniFilesTest : BaseTest() {
+
+    @Test
+    fun gettingStartedWithINIConfigfiles() {
```
Comment (oxisto) [inline cpg-language-configfiles/src/test/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesTest.kt:46]: although not necessary, all other tests start with `test`

File: cpg-language-configfiles/src/test/resources/config.ini [hunk 10]
```diff
@@ -0,0 +1,12 @@
+; Wikipedia example https://en.wikipedia.org/wiki/INI_file
```
Comment (oxisto) [inline cpg-language-configfiles/src/test/resources/config.ini:1]: what's the license of the wikipedia file? Maybe just better to have our own example?

File: build.gradle.kts [hunk 11]
```diff
@@ -143,3 +143,9 @@ val enableJVMFrontend: Boolean by extra {
     enableJVMFrontend.toBoolean()
 }
 project.logger.lifecycle("JVM frontend is ${if (enableJVMFrontend) "enabled" else "disabled"}")
+
+val enableConfigfilesFrontend: Boolean by extra {
```
Comment (oxisto) [inline build.gradle.kts:147]: -> `enableConfigFilesFrontend` ?

File: cpg-language-configfiles/build.gradle.kts (hunk 6 above)
Comment (oxisto) [reply cpg-language-configfiles/build.gradle.kts:32]: Or singular as well `cpg-language-configfile` ? or just `cpg-language-config` or specific `cpg-langauge-ini`? not quite sure if we want to extend this specific frontend to ALL config files?

Comment (maximiliankaul) [reply cpg-language-configfiles/build.gradle.kts]: My intention was to collect all config file frontends here. However, I changed this as per your review.

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt (hunk 7 above)
Comment (maximiliankaul) [reply cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesFrontend.kt]: Done

File: cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt (hunk 8 above)
Comment (maximiliankaul) [reply cpg-language-configfiles/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesLanguage.kt]: Done

File: cpg-language-configfiles/src/test/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesTest.kt (hunk 9 above)
Comment (maximiliankaul) [reply cpg-language-configfiles/src/test/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFilesTest.kt]: Done

File: cpg-language-configfiles/src/test/resources/config.ini (hunk 10 above)
Comment (maximiliankaul) [reply cpg-language-configfiles/src/test/resources/config.ini]: Good point. Done.

File: build.gradle.kts (hunk 11 above)
Comment (maximiliankaul) [reply build.gradle.kts]: I changed this to `enableINIFrontend` to be in line with the review changes above.

File: cpg-language-ini/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFileFrontend.kt [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        } catch (ex: Exception) {
+            throw TranslationException("Parsing failed with exception: $ex")
+        }
+
+        /*
+         * build a namespace name relative to the configured
+         * [de.fraunhofer.aisec.cpg.TranslationConfiguration.topLevel] using
+         * [Language.namespaceDelimiter] as a separator
+         */
+        val topLevel = config.topLevel ?: file.parent
```
Comment (maximiliankaul) [inline cpg-language-ini/src/main/kotlin/de/fraunhofer/aisec/cpg/frontend/configfiles/IniFileFrontend.kt:92]: @KuechA will resolve this ;)

**Review by oxisto — APPROVED** [review]

**Pull request merged**